Eliminate TypeVars and Wildcards early in glb/lub

joroKr21 opened the pull request:
Type variables and wildcards can act both as subtypes and
supertypes depending on the direction of the test. This means
types with variables and/or wildcards can shadow other types in
`glb/lub`.

Consider the following examples:
```scala
lub((Int, ? ) :: (Int, String) :: Nil)
glb((Int, ?A) :: (Int, String) :: Nil)
```

Intuitively both should return `(Int, String)`, but this depends
on the order of types in the list passed to `glb/lub`.
This commit incorporates the simple heuristic of moving types that
contain variables and/or wildcards to the end of the list.

Fixes scala/bug#10686 and fixes scala/bug#10740

joroKr21 commented:
I'm not sure how this would affect performance

joroKr21 commented:
Is it possible that there is a deadlock somewhere in `concurrent.TrieMap` that is sometimes triggered by the scalacheck tests?

joroKr21 commented:
/rebuild

joroKr21 commented:
@retronym how does this look?

joroKr21 commented on src/reflect/scala/reflect/internal/tpe/GlbLubs.scala:
```diff
@@ -20,6 +20,10 @@ private[internal] trait GlbLubs {
   /** In case anyone wants to turn off lub verification without reverting anything. */
   private final val verifyLubs = true
 
+  private object wildcardExists extends FindTypeCollector(_.isWildcard) {
```
Should I check for `isGround` instead? I'm not sure.

retronym commented:
To help me understand your patch, I've [attempted to split](https://github.com/scala/scala/compare/2.13.x...retronym:review/6257?expand=1) it into a refactoring and a behavior change.

Does that split look faithful to your intent?

I still need a bit more time to understand the change.

joroKr21 commented:
Your split is almost correct. Apologies for not doing this myself.

The semantic change is just the addition of this switch of the current type from `t` to `t1`:
```scala
if (lteq(t, t1) && t.exists(_.isWildcard)) t1 else t
```
Note that if we're not going to check for wildcards the switch is pointless and there would be no need to compare `lteq(t, t1)`.

The (tail)-recursive version of the "find all maximal elements in a list given a partial order" (here `maxTypes`) made it easier for me to understand what it's actually doing. Here the worst running time is inherently O(n^2). There is a difference in the order that types are compared. In `maxTypes` if we find a bigger element than the current one, we immediately switch to it and go back to the beginning of the list. The old implementation always uses the head of the list to eliminate smaller elements and then recurses into the tail.

joroKr21 commented:
Hold on, it looks to me that `maxTypes` doesn't do enough comparisons, so there must be a bug.
The most annoying part is to preserve the original order of the list. This shouldn't matter right?

joroKr21 commented:
Should be good now

joroKr21 commented on src/reflect/scala/reflect/internal/tpe/GlbLubs.scala:
```diff
@@ -182,31 +182,46 @@ private[internal] trait GlbLubs {
+        @tailrec def inner(work: List[Type], done: List[Type]): List[Type] = work match {
+          case t2 :: ts2 if po(t1, t2) => // found a bigger type, we're done here
+            if (!t2.exists(_.isWildcard) || !po(t2, t1)) max
+            else t1 :: (ts2 reverse_::: done)
```
Could also do `ts2 reverse_::: t1 :: done` here, but it doesn't seem to matter.

joroKr21 replied:
Huh, I guess it does matter after all - so frustrating that `with` is not symmetric.

joroKr21 committed:
Factor out duplicated logic from elimSub/elimSuper

joroKr21 committed:
Eliminate TypeVars and Wildcards early in glb/lub

Type variables and wildcards can act both as subtypes and
supertypes depending on the direction of the test. This means
types with variables and/or wildcards can shadow other types in
`glb/lub`.

Consider the following examples:
```scala
lub((Int, ? ) :: (Int, String) :: Nil)
glb((Int, ?A) :: (Int, String) :: Nil)
```

Intuitively both should return `(Int, String)`, but this depends
on the order of types in the list passed to `glb/lub`.
This commit incorporates the simple heuristic of moving types that
contain variables and/or wildcards to the end of the list.

Fixes scala/bug#10686 and fixes scala/bug#10740

joroKr21 commented:
@retronym It looks like the list passed to `lub`/`glb` can contain `TypeVar`s which depend on the exact order of comparisons as being called in the current implementation :sob: so I can't change it in any way whatsoever. All I can do is push the `partition` on `?` inside.

joroKr21 commented:
/sync

joroKr21 commented:
Updated to fix scala/bug#10740

An alternative would be to use a `Queue` for type constraints and prepend / append as needed.

Note that this also compiles in Dotty, but neither on Scala 2.13.x nor with this PR:
```scala
def foo[A, B](implicit ev: (Int, A) =:= (String, B)) = 42
foo
```

This would require a redesign of how type variables and constraints are handled.

SethTisue commented:
/synch

joroKr21 commented:
/sync

joroKr21 commented:
/sync

adriaanm commented on src/reflect/scala/reflect/internal/tpe/GlbLubs.scala:
```diff
@@ -158,31 +158,30 @@ private[internal] trait GlbLubs {
 
+  /** From a list of types, retain only maximal types as determined by the partial order `po`. */
+  private def maxTypes(ts: List[Type])(po: (Type, Type) => Boolean): List[Type] = {
+    def loop(ts: List[Type]): List[Type] = ts match {
```
`@tailrec`?

adriaanm commented on src/reflect/scala/reflect/internal/tpe/GlbLubs.scala:
```diff
@@ -158,31 +158,30 @@ private[internal] trait GlbLubs {
+      case t :: ts1 =>
+        val ts2 = loop(ts1.filterNot(po(_, t)))
+        if (ts2.exists(po(t, _))) ts2 else t :: ts2
+      case Nil => Nil
```
I assume we special-cased `Nil` and the singleton list based on profiling before? Before we drop these cases, let's gather some statistics on the frequency with which they occur (from casual inspection, not all callers ensure the list's length is >= 2).

joroKr21 replied:
It would be a good exercise for me - any pointers how to do this?

adriaanm replied:
Check out `Statistics.incCounter`

adriaanm commented on src/reflect/scala/reflect/internal/tpe/GlbLubs.scala:
```diff
@@ -167,7 +167,14 @@ private[internal] trait GlbLubs {
-    loop(ts)
+    // The order here matters because type variables and
+    // wildcards can act both as subtypes and supertypes.
+    val (ts2, ts1) = ts.partition(_ exists {
```
I don't think this is the right solution. We shouldn't be calling `lub`/`glb` with a list of types that includes type variables or wildcards. These test cases are a great help in debugging this.

joroKr21 replied:
Any idea how we could avoid that?

joroKr21 replied:
Recalling my experiments here, the current tests exercise this code path, i.e. type variables being inferred by virtue of participating in the `lub/glb`.

The complete solution would probably come with Dotty but in the mean time it's unsatisfactory explaining to people that they have to swap the arguments to `=:=` to make their code compile :smile: (due to scala/bug#10686). How about approximating just the wildcards?

adriaanm replied:
I agree we should improve, but I'd like to take a look at the actual code paths that lead here, and how to remove those non-sensical types from the beginning. The type checker already approximates the expected type using wildcards (as `wildPt`), which shares a similar motiviation.

joroKr21 replied:
One code path is via `ExistentialType.withTypeVars` in subtyping checks when the other type also has type vars. Try this and see a few of the pos partests fail with "trying to do lub/glb of typevar":

```diff
--- a/src/reflect/scala/reflect/internal/tpe/TypeComparers.scala
+++ b/src/reflect/scala/reflect/internal/tpe/TypeComparers.scala
@@ -529,7 +529,7 @@ trait TypeComparers {
         (rt2.parents forall (isSubType(tp1, _, depth))) &&
           (rt2.decls forall (specializesSym(tp1, _, depth)))
       case et2: ExistentialType =>
-        et2.withTypeVars(isSubType(tp1, _, depth), depth) || fourthTry
+        suspendingTypeVars(typeVarsInType(tp1))(et2.withTypeVars(isSubType(tp1, _, depth), depth)) || fourthTry
       case mt2: MethodType =>
         tp1 match {
           case mt1 @ MethodType(params1, res1) =>
```

I'm not sure what is the correct thing to do here. Since we only care if there is any solution to the existential skolems but not what that solution is, if feels unexpected that this would affect the inference of some unrelated type variables.

adriaanm replied:
Agreed that those unrelated type vars should not be constrained -- they should act like skolems, I guess. I'm reluctant to change this until I fully understand how this works, which is not likely to happen anymore for 2.13.0, sadly.

adriaanm replied:
My initial comment was wrong -- I see now from your ltr/rtl test case that we get into the situation with wildcards in types. I can't think of a better way to make this more consistent, so, let's go with yours.

adriaanm commented:
Also, I was poking around here and was reminded of `stripExistentialsAndTypeVars`. Maybe that method is the right target for this PR?

joroKr21 commented:
In its current form it doesn't help, because it handles only top-level type variables and it just bails out, but I will try out something similar with a `TypeMap`.

adriaanm commented:
Yes, I meant this PR could "target" that method so that it also the general case. Maybe it was a mistake that it only handles top-level ones?

adriaanm reviewed (approved):
Agreed this is a problem, and I can't think of a better way to solve it. So, let's book the progress. Would like to make sure performance is not affected before merging /cc @retronym.

adriaanm commented:
Since we have a lot going on, optimistically merging -- if performance is affected, we can fix before RC1.
